### DISCUSSION TITLE
Theme setting not rendering when the server is initially loaded.

### DISCUSSION THREAD

**Discussion by thibpun** [abereghici/remix-themes#44] [Q&A] [answered]
I've imported the library as provided in the docs but it doesn't render properly.  Here is my root.tsx code:
```
import {
  Links,
  Meta,
  Outlet,
  Scripts,
  ScrollRestoration,
  useLoaderData,
} from "@remix-run/react";
import "./tailwind.css";
import NavBar from "./components/NavBar";
import { LoaderFunctionArgs, ActionFunctionArgs } from "@remix-run/node";
import authenticator from "utils/auth.server";
import {
  ThemeProvider,
  useTheme,
  PreventFlashOnWrongTheme,
} from "remix-themes/build/theme-provider";
import { themeSessionResolver } from "utils/theme.server";

export async function loader({ request }: LoaderFunctionArgs) {
  const user = await authenticator.isAuthenticated(request);
  const { getTheme } = await themeSessionResolver(request);
  return {
    user: user,
    theme: getTheme(),
  };
}

export async function action({ request }: ActionFunctionArgs) {
  return await authenticator.logout(request, {
    redirectTo: "/",
  });
}

function Layout({ children }: { children: React.ReactNode }) {
  const loaderData = useLoaderData<typeof loader>();
  const [theme, setTheme] = useTheme();
  console.log(theme);
  return (
    <html lang="en" className={theme ?? ""}>
      <head>
        <meta charSet="utf-8" />
        <meta name="viewport" content="width=device-width, initial-scale=1" />
        <Meta />
        <Links />
      </head>
      <body className="bg-background text-text" suppressHydrationWarning>
        <header>
          <NavBar />
        </header>
        {children}
        <ScrollRestoration />
        <PreventFlashOnWrongTheme ssrTheme={Boolean(loaderData?.theme)} />
        <Scripts />
      </body>
    </html>
  );
}

function Providers({ children }: { children: React.ReactNode }) {
  const loaderData = useLoaderData<typeof loader>();
  return (
    <ThemeProvider
      specifiedTheme={loaderData.theme}
      themeAction="/action/set-theme"
    >
      <Layout>{children}</Layout>
    </ThemeProvider>
  );
}

export default function App() {
  return (
    <Providers>
      <Outlet />
    </Providers>
  );
}

```
The ThemeProvider gets the system theme and sets it correctly in the className attribute, But the app is stuck in light mode even though the <html> className is "dark". In another file I use the setTheme function, it changes the theme in the cookie and the className, again nothing happens in the render.

I've found out that  If I change the html tag from: ``<html lang="en" className={theme ?? ""}>`` 
to: ``<html lang="en" className={theme ?? "dark"}>``, then the app works as intended.
And even if I change it back to ``<html lang="en" className={theme ?? ""}>`` the app will work as intended until I restart the server.

But then I found that I can instead add:
```
// if (theme !== "light") temp = "dark";
```
so the layout component looks like this:
```
function Layout({ children }: { children: React.ReactNode }) {
... omitted for brevity
// if (theme !== "light") temp = "dark";
return (
    <html lang="en" className={theme ?? ""}>
... omitted for brevity
```
the app works, **and yes the code I added is commented out and the temp variable is neither declared anywhere nor is it called anywhere else**.

if instead I add:
```
  if (theme !== "light"){}

```
the same if statement this time not commented out but the body of the statement is empty the app doesn't work again.

If I uncomment and instantiate the temp variable the app works:
``` 
let temp = "";
  if (theme !== "light") {
    temp = "dark";
  }
```
If then I delete the if condition the app doesn't work again:
``` 
let temp = "";
```
But then when I write it like this:
``` 
let temp = "dark";
```
the app works again. And I **stress that the temp variable is not used anywhere in the app it's declared and discarded in the same line of code**.

 So lastly I can also make the app work writing only: 
```
("dark");
```
as a string or:
```
//dark
```
as a comment. 
And I don't need to add this to the root, I can create a new route/component whose **only** line is ```//dark``` and the app will work, this route/component is not called anywhere.

Any other word like "light"/"pink"/"bicycle" doesn't work, nor do the permutations of the word dark work eg. dakr. Also capitalizing Dark doesn't work.

In the end it would seem that all I needed to do was add the word: dark ***anywhere*** In my project and the bug would have been fixed and I wouldn't have spent 3hrs on debugging. But at this point I just need this to make sense, does me adding the word "dark" send a fetch request to the Universe which responds by sending cosmic rays which flip the bits on my server making it work or is there a comprehensible answer?

**Comment by abereghici** [comment]
Hey. Could you please try this setup? 
```
import { json, LoaderFunctionArgs } from "@remix-run/node";
import {
  Links,
  Meta,
  Outlet,
  Scripts,
  ScrollRestoration,
  useRouteLoaderData,
} from "@remix-run/react";
import {
  ThemeProvider,
  useTheme,
  PreventFlashOnWrongTheme,
  Theme,
} from "remix-themes";
import { themeSessionResolver } from "utils/theme.server";

export async function loader({ request }: LoaderFunctionArgs) {
  const { getTheme } = await themeSessionResolver(request);
  return json({ theme: getTheme() });
}

export function Layout({ children }: { children: React.ReactNode }) {
  const data = useRouteLoaderData<typeof loader>("root");

  return (
    <ThemeProvider specifiedTheme={data?.theme as Theme} themeAction="/action/set-theme">
      <InnerLayout ssrTheme={Boolean(data?.theme)}>{children}</InnerLayout>
    </ThemeProvider>
  );
}

function InnerLayout({
  ssrTheme,
  children,
}: {
  ssrTheme: boolean;
  children: React.ReactNode;
}) {
  const [theme] = useTheme();

  return (
    <html suppressHydrationWarning lang="en" className={theme ?? ""}>
      <head>
        <meta charSet="utf-8" />
        <meta name="viewport" content="width=device-width, initial-scale=1" />
        <Meta />
        <Links />
      </head>
      <body>
        {children}
        <ScrollRestoration />
        <PreventFlashOnWrongTheme ssrTheme={ssrTheme} />
        <Scripts />
      </body>
    </html>
  );
}

export default function App() {
  return <Outlet />;
}

```

**Reply by thibpun** [reply]
Tried it, the same issue persists.

**Comment by abereghici** [comment]
@thibpun could you please create a dummy project with your theme code setup and share here? Yesterday just started a new project and I added the library and I didn't encounter any issues.

**Reply by thibpun** [reply]
Here's the repo:https://github.com/thibpun/remix-themes-bug

**Comment by abereghici** [comment, marked as answer]
@thibpun thank you! I created a pull request with a fix here https://github.com/thibpun/remix-themes-bug/pull/1 
The issue was related to tailwind.

**Reply by thibpun** [reply]
Thank you. I thought it could be related to tailwind but the app worked when I would use my own theme variable which could be set to "dark" or "" using setState. 
One more question. You removed the useEffect from the root.tsx, I used it to "refresh" the cookie so it wouldn't expire do you have a better way I could do it?

**Reply by abereghici** [reply]
I wouldn't worry about it, but if you want, you can refresh the cookie from your `loader`. 

```
export async function loader({ request }: LoaderFunctionArgs) {
  const user = await authenticator.isAuthenticated(request);
  const { setTheme, getTheme, commit } = await themeSessionResolver(request);
  
  const theme = getTheme();
  setTheme(theme);

  return json({
    user: user,
    theme: getTheme(),
  }, 
   {
        headers: {'Set-Cookie': await session.commit()},
  });
}
```

I didn't try this code, but it's a start for you to check this path.
